lantern_embeddings/src/lib.rs

```diff
@@ -0,0 +1,282 @@
+            .query(
+                &format!("SELECT COUNT(\"{}\") FROM \"{}\";", pk, table),
+                &[],
+            )
+            .unwrap();
+        let count: i64 = rows[0].get(0);
+        println!("[*] Found {} items in table \"{}\"", count, table);
+        // With portal we can execute a query and poll values from it in chunks
+        let portal = transaction
+            .bind(
+                &format!("SELECT \"{}\", \"{}\" FROM \"{}\";", pk, column, table),
+                &[],
+            )
+            .unwrap();
+
+        loop {
+            // poll batch_size rows from portal and send it to embedding thread via channel
+            let rows = transaction
+                .query_portal(&portal, batch_size as i32)
+                .unwrap();
+
+            if rows.len() == 0 {
+                break;
+            }
+
+            if tx.send(rows).is_err() {
+                break;
+            }
+        }
+        drop(tx);
+    });
+
+    return Ok(handle);
+}
+
+fn embedding_worker(
+    args: &cli::EmbeddingArgs,
+    rx: Receiver<Vec<Row>>,
+    tx: Sender<Vec<EmbeddingRecord>>,
+) -> Result<JoinHandle<()>, anyhow::Error> {
+    let is_visual = args.visual.clone();
+    let model = args.model.clone();
+    let data_path = args.data_path.clone();
+    let mut count: u64 = 0;
+
+    let handle = std::thread::spawn(move || {
+        let mut start = Instant::now();
+        loop {
+            let rows = rx.recv();
+            if rows.is_err() {
+                // channel has been closed
+                break;
+            }
+
+            if count == 0 {
+                // mark exact start time
+                start = Instant::now();
+            }
+
+            let rows = rows.unwrap();
+            let mut input_vectors = Vec::with_capacity(rows.len());
+
+            for row in &rows {
+                let col: &str = row.get(1);
+                input_vectors.push(col);
+            }
+
+            let response_embeddings = if is_visual {
+                clip::process_image(&model, &input_vectors, None, data_path.as_deref())
+            } else {
+                clip::process_text(&model, &input_vectors, None, data_path.as_deref())
+            };
+
+            if let Err(e) = response_embeddings {
+                panic!("{}", e);
+            }
+
+            let response_embeddings = response_embeddings.unwrap();
+
+            count += response_embeddings.len() as u64;
+
+            let duration = start.elapsed().as_secs();
+            // avoid division by zero error
+            let duration = if duration > 0 { duration } else { 1 };
+            println!(
+                "[*] Generated {} embeddings - speed {} emb/s",
+                count,
+                count / duration
+            );
+
+            let mut response_data = Vec::with_capacity(rows.len());
+
+            for (i, embedding) in response_embeddings.iter().enumerate() {
+                let id: i32 = rows[i].get(0);
```

Comment:
Note that the primary key passed on the table may not be an integer (uuid is a common non-integer pk).
You can either not support it or handle it accordingly via generics or otherwise